src/server/tablet_snap.rs

```diff
@@ -80,21 +84,57 @@ fn is_sst(file_name: &str) -> bool {
     file_name.ends_with(".sst")
 }
 
-async fn read_to(f: &mut File, to: &mut Vec<u8>, size: usize, limiter: &Limiter) -> Result<()> {
-    // It's likely in page cache already.
-    let cost = size / 2;
-    limiter.consume(cost).await;
-    SNAP_LIMIT_TRANSPORT_BYTES_COUNTER_STATIC
-        .send
-        .inc_by(cost as u64);
-    to.clear();
-    to.reserve_exact(size);
-    let mut buf: BorrowedBuf<'_> = to.spare_capacity_mut().into();
-    f.read_buf_exact(buf.unfilled())?;
-    unsafe {
-        to.set_len(size);
+struct WrappedReadableFile(Either<File, encryption_export::DecrypterReader<File>>);
+
+impl WrappedReadableFile {
+    fn open(key_manager: &Option<Arc<DataKeyManager>>, path: &Path) -> Result<Self> {
+        let f = File::open(path)?;
+        let inner = if let Some(m) = key_manager {
+            Either::Right(
+                m.open_file_with_reader(path, f)
+                    .map_err(|e| Error::Other(e.into()))?,
+            )
+        } else {
+            Either::Left(f)
+        };
+        Ok(Self(inner))
+    }
+
+    fn seek(&mut self, to: SeekFrom) -> Result<u64> {
+        let r = match &mut self.0 {
+            Either::Left(f) => f.seek(to)?,
+            Either::Right(f) => f.seek(to)?,
+        };
+        Ok(r)
+    }
+
+    fn len(&self) -> Result<u64> {
+        let r = match &self.0 {
+            Either::Left(f) => f.metadata()?.len(),
+            Either::Right(f) => f.inner().metadata()?.len(),
+        };
+        Ok(r)
+    }
+
+    async fn read_to(&mut self, to: &mut Vec<u8>, size: usize, limiter: &Limiter) -> Result<()> {
+        // It's likely in page cache already.
+        let cost = size / 2;
+        limiter.consume(cost).await;
+        SNAP_LIMIT_TRANSPORT_BYTES_COUNTER_STATIC
+            .send
+            .inc_by(cost as u64);
+        to.clear();
+        to.reserve_exact(size);
+        let mut buf: BorrowedBuf<'_> = to.spare_capacity_mut().into();
+        match &mut self.0 {
+            Either::Left(f) => f.read_buf_exact(buf.unfilled())?,
+            Either::Right(f) => f.read_buf_exact(buf.unfilled())?,
```

Comment:
Why decrypt the file?

Reply:
Strictly speaking it's possible that different plaintext being encrypted with different keys still output the same results. So checking the plaintext is more correct.